fem/bilininteg_dgtrace.cpp

```diff
@@ -0,0 +1,1089 @@
+                             const double alpha,
+                             const double beta,
+                             Vector &op)
+{
+   const int VDIM = 3;
+
+   auto d = Reshape(det.Read(), Q1D, Q1D, NF);
+   auto n = Reshape(nor.Read(), Q1D, Q1D, VDIM, NF);
+   const bool const_r = rho.Size() == 1;
+   auto R =
+      const_r ? Reshape(rho.Read(), 1,1,1) : Reshape(rho.Read(), Q1D,Q1D,NF);
+   const bool const_v = vel.Size() == 3;
+   auto V =
+      const_v ? Reshape(vel.Read(), 3,1,1,1) : Reshape(vel.Read(), 3,Q1D,Q1D,NF);
+   auto W = w.Read();
+   auto qd = Reshape(op.Write(), Q1D, Q1D, 2, 2, NF);
+
+   MFEM_FORALL(f, NF,//can be optimized with Q1D*Q1D threads for NF blocks
+   {
+      for (int q1 = 0; q1 < Q1D; ++q1)
+      {
+         for (int q2 = 0; q2 < Q1D; ++q2)
+         {
+            const double r = const_r ? R(0,0,0) : R(q1,q2,f);
+            const double v0 = const_v ? V(0,0,0,0) : V(0,q1,q2,f);
+            const double v1 = const_v ? V(1,0,0,0) : V(1,q1,q2,f);
+            const double v2 = const_v ? V(2,0,0,0) : V(2,q1,q2,f);
+            const double dot = n(q1,q2,0,f) * v0 + n(q1,q2,1,f) * v1 + n(q1,q2,2,f) * v2;
+            const double abs = dot > 0.0 ? dot : -dot;
+            const double w = W[q1+q2*Q1D]*r*d(q1,q2,f);
+            qd(q1,q2,0,0,f) = w*( alpha/2 * dot + beta * abs );
+            qd(q1,q2,1,0,f) = w*( alpha/2 * dot - beta * abs );
+            qd(q1,q2,0,1,f) = w*(-alpha/2 * dot - beta * abs );
+            qd(q1,q2,1,1,f) = w*(-alpha/2 * dot + beta * abs );
+         }
+      }
+   });
+}
+
+static void PADGTraceSetup(const int dim,
+                           const int D1D,
+                           const int Q1D,
+                           const int NF,
+                           const Array<double> &W,
+                           const Vector &det,
+                           const Vector &nor,
+                           const Vector &rho,
+                           const Vector &u,
+                           const double alpha,
+                           const double beta,
+                           Vector &op)
+{
+   if (dim == 1) { MFEM_ABORT("dim==1 not supported in PADGTraceSetup"); }
+   if (dim == 2)
+   {
+      PADGTraceSetup2D(Q1D, NF, W, det, nor, rho, u, alpha, beta, op);
+   }
+   if (dim == 3)
+   {
+      PADGTraceSetup3D(Q1D, NF, W, det, nor, rho, u, alpha, beta, op);
+   }
+}
+
+void DGTraceIntegrator::SetupPA(const FiniteElementSpace &fes, FaceType type)
+{
+   nf = fes.GetNFbyType(type);
+   if (nf==0) { return; }
+   // Assumes tensor-product elements
+   Mesh *mesh = fes.GetMesh();
+   const FiniteElement &el = *fes.GetTraceElement(0,
+                                                  fes.GetMesh()->GetFaceBaseGeometry(0));
+   FaceElementTransformations &T = *fes.GetMesh()->GetFaceElementTransformations(
+                                      0);
+   const IntegrationRule *ir = &GetRule(el.GetGeomType(), el.GetOrder(), T);
```

Comment:
If `this->IntRule` is not `NULL`, we should use that, e.g.:
```c++
    const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el.GetGeomType(), el.GetOrder(), T);
```

Reply:
done.